transportable-udfs-annotation-processor/src/main/java/com/linkedin/transport/processor/TransportProcessor.java

```diff
@@ -0,0 +1,189 @@
+ *
+ * The annotation processor does not rely on a special annotation, instead it will look for non-abstract classes which
+ * extend {@link StdUDF} and implement {@link TopLevelStdUDF}. It will then identify all overloadings of the same UDF
+ * name and store them in the resource file.
+ */
+@SupportedOptions({"debug"})
+@SupportedAnnotationTypes({"*"})
+public class TransportProcessor extends AbstractProcessor {
+
+  private Types _types;
+  private Elements _elements;
+  private TypeMirror _topLevelStdUDFInterfaceType;
+  private TypeMirror _stdUDFClassType;
+  private UDFProperties _udfProperties;
+
+  @Override
+  public SourceVersion getSupportedSourceVersion() {
+    return SourceVersion.latestSupported();
+  }
+
+  @Override
+  public synchronized void init(ProcessingEnvironment processingEnv) {
+    super.init(processingEnv);
+    _types = processingEnv.getTypeUtils();
+    _elements = processingEnv.getElementUtils();
+    _topLevelStdUDFInterfaceType = _elements.getTypeElement(TopLevelStdUDF.class.getName()).asType();
+    _stdUDFClassType = _elements.getTypeElement(StdUDF.class.getName()).asType();
+    _udfProperties = new UDFProperties();
+  }
+
+  @Override
+  public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
+    try {
+      processImpl(roundEnv);
+    } catch (Exception e) {
+      // We don't allow exceptions of any kind to propagate to the compiler
+      try (StringWriter stringWriter = new StringWriter(); PrintWriter printWriter = new PrintWriter(stringWriter)) {
+        e.printStackTrace(printWriter);
+        fatalError(stringWriter.toString());
+      } catch (IOException ioe) {
+        fatalError("Could not close resources " + ioe);
+      }
+    }
+    // Universal processors should return false since other processor can be potentially acting on the same element
+    return false;
+  }
+
+  private void processImpl(RoundEnvironment roundEnv) {
+    if (roundEnv.processingOver()) {
+      generateUDFPropertiesFile();
+    } else {
+      processElements(roundEnv.getRootElements());
+    }
+  }
+
+  private void processElements(Set<? extends Element> elements) {
+    for (Element element : elements) {
+      // Check if the element is a non-abstract class which extends the StdUDF class
+      if (element.getKind().equals(ElementKind.CLASS) && !element.getModifiers().contains(Modifier.ABSTRACT)
+          && _types.isAssignable(element.asType(), _stdUDFClassType)) {
+        processUDFClass((TypeElement) element);
+      }
+    }
+  }
+
+  /**
+   * Finds the {@link TopLevelStdUDF} for the given {@link StdUDF} class and adds it to the list of discovered UDFs
+   */
+  private void processUDFClass(TypeElement udfClassElement) {
+    debug(String.format("Processing UDF Class: %s", udfClassElement.getQualifiedName()));
+
+    TypeElement closestTopLevelStdUDFInterface = getClosestTopLevelStdUDFInterface(udfClassElement);
+    if (closestTopLevelStdUDFInterface != null) {
```

Comment:
getClosestTopLevelStdUDFInterface both logs an error and returns null. I think it is more reasonable to return null only there and log an error here. Otherwise, the else branch of this if statement looks missing.

Reply:
Please verify if this comment is still valid after the refactor